proxy/ParentYamlConfig.cc

```diff
@@ -0,0 +1,224 @@
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+ */
+
+#include <memory>
+#include <vector>
+
+#include "ParentYamlConfig.h"
+
+ParentYamlConfig::ParentYamlConfig() : dest_size(0) {}
+
+ParentYamlConfig::~ParentYamlConfig() {}
+
+parent_dest &
+operator>>(const YAML::Node &node, parent_dest &p)
+{
+  // get all the strings
+  if (node["dest_domain"]) {
+    p.dest_domain = node["dest_domain"].as<std::string>();
+  }
+  if (node["dest_host"]) {
+    p.dest_host = node["dest_host"].as<std::string>();
+  }
+  if (node["dest_ip"]) {
+    p.dest_ip = node["dest_ip"].as<std::string>();
+  }
+  if (node["url_regex"]) {
+    p.url_regex = node["url_regex"].as<std::string>();
+  }
+  if (node["scheme"]) {
+    p.scheme = node["scheme"].as<std::string>();
+  }
+  if (node["prefix"]) {
+    p.prefix = node["prefix"].as<std::string>();
+  }
+  if (node["suffix"]) {
+    p.suffix = node["suffix"].as<std::string>();
+  }
+  if (node["method"]) {
+    p.method = node["method"].as<std::string>();
+  }
+  if (node["time"]) {
+    p.time = node["time"].as<std::string>();
+  }
+  if (node["src_ip"]) {
+    p.src_ip = node["src_ip"].as<std::string>();
+  }
+  if (node["parent"]) {
+    p.parent = node["parent"].as<std::string>();
+  }
+  if (node["secondary_parent"]) {
+    p.secondary_parent = node["secondary_parent"].as<std::string>();
+  }
+  if (node["parent_retry"]) {
+    p.parent_retry = node["parent_retry"].as<std::string>();
+  }
+  if (node["unavailable_server_retry_responses"]) {
+    p.unavailable_server_retry_responses = node["unavailable_server_retry_responses"].as<std::string>();
+  }
+  if (node["round_robin"]) {
+    p.round_robin = node["round_robin"].as<std::string>();
+  }
+  if (node["qstring"]) {
+    p.qstring = node["qstring"].as<std::string>();
+  }
+
+  // get all the int's
+  if (node["port"]) {
+    p.port = node["port"].as<int>();
+  }
+  if (node["secondary_mode"]) {
+    p.secondary_mode = node["secondary_mode"].as<int>();
+  }
+  if (node["max_simple_retries"]) {
+    p.max_simple_retries = node["max_simple_retries"].as<int>();
+  }
+  if (node["max_unavailable_server_retries"]) {
+    p.max_unavailable_server_retries = node["max_unavailable_server_retries"].as<int>();
+  }
+
+  // get all the booleans
+  if (node["internal"]) {
+    p.internal = node["internal"].as<bool>();
+  }
+  if (node["parent_is_proxy"]) {
+    p.parent_is_proxy = node["parent_is_proxy"].as<bool>();
+  }
+  if (node["go_direct"]) {
+    p.go_direct = node["go_direct"].as<bool>();
+  }
+
+  return p;
+}
+
+int
+ParentYamlConfig::Init(const std::string &content)
+{
+  YAML::Node root;
+  YAML::Node destinations;
+
+  try {
+    root = YAML::Load(content);
+
+    if (!root.IsMap()) {
+      error_msg << __FILE__ << ":" << __LINE__ << ": "
+                << "yaml parse error, expected a map at the root node.";
```

Comment:
Nitpick:  you can write error_msg << * __FILE__* ":" because __FILE__ expands to a string literal and the compiler will concatenate string literals.

In any case, I'm not sure how helpful source file and line is here.  Seems worth it to pass in the name of the file that's being parsed, to include it in error messages.

Reply:
resolved, not using this.